### PR TITLE
files: Allow extra headers for file download in DownloadHeaders

### PR DESCRIPTION
**Pull request by ncw**
~~Note that this is based off my apiImpl branch - will rebase as required.~~

Now a standalone branch.

### CODE REVIEW

**Comment by diwakergupta** [comment]
Thanks for the PR @ncw! Unfortunately this approach won't work as `client.go` is auto-generated and so will get overwritten the next time I update the SDK. We'll either have to update the generator itself, or explore alternatives. WDYT?

**Comment by ncw** [comment]
I'd be happy to have a go at updating the generator if you think that is the right approach.

If we're updating the generator then the sensible thing would be to put the optional headers in the `DownloadArg` I would have thought, which will have minimal API impact elsewhere.

I'll have a poke about at it and see what I can come up with!

**Comment by ncw** [comment]
OK I've done a patch for the generator.

There is probably a more elegant way of selecting the things which need patching other than by name though. `style == "download"` looked promising but I couldn't work out how to read that in the types generator.

Let me know what you think!

**Comment by ncw** [comment]
Any thoughts on this approach?

**Comment by ncw** [comment]
Any thoughts on this approach?  I'd like to un-fork this repo if possible!

Thanks

Nick

File: generator/generate-sdk.sh [hunk 1]
```diff
@@ -20,6 +20,6 @@ pushd ${spec_dir}
 spec_version=$(git rev-parse --short HEAD)
 popd
 
-sed -i '' -e "s/UNKNOWN SDK VERSION/${sdk_version}/" \
+sed -i  -e "s/UNKNOWN SDK VERSION/${sdk_version}/" \
```
Comment (diwakergupta) [inline generator/generate-sdk.sh]: Please submit unrelated changes in a separate PR, thanks!

**Comment by diwakergupta** [comment]
@ncw sorry about the delay here. There was some recent changes that might offer an alternative (simpler) approach. Have you seen `config.HeaderGenerator`? https://github.com/dropbox/dropbox-sdk-go-unofficial/blob/master/dropbox/sdk.go#L59

Check it out and LMK if you can use that instead.

**Comment by ncw** [comment]
@diwakergupta config.HeaderGenerator looks useful, but not quite useful enough I think.  The trouble is that you would typically have only one Config struct so changing HeaderGenerator on the fly is troublesome in a highly concurrent program like rclone.

What is required is the ability to do this on a per transaction basis.

I'll split the first two commits in this PR into two separate PRs in a moment.

**Comment by ncw** [comment]
I've rebased this commit and split the other two changes into their own PRs - hopefully that makes things a bit easier to understand.

File: generator/go_client.stoneg.py [hunk 2]
```diff
@@ -90,7 +90,10 @@ def _generate_route(self, namespace, route):
             self._generate_request(namespace, route)
             self._generate_post()
             self._generate_response(route)
-            with self.block('if resp.StatusCode == http.StatusOK'):
+            ok_check = 'if resp.StatusCode == http.StatusOK'
+            if fn == "Download":
```
Comment (diwakergupta) [inline generator/go_client.stoneg.py]: Rather than check on the function name, do you want to predicate this on the "style" attribute instead? That might be a bit more generic. Otherwise this diff LGTM!

Comment (diwakergupta) [inline generator/go_client.stoneg.py]: @ncw ICYMI ^ LMK, otherwise I can merge this as-is and we can improve things in a separate commit.
